File: scratch/find_duplicates.mjs
import fs from 'fs';
import path from 'path';

function getMdxFiles(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        if (stat && stat.isDirectory()) {
            results = results.concat(getMdxFiles(filePath));
        } else if (file.endsWith('.mdx')) {
            results.push(filePath);
        }
    });
    return results;
}

function cleanText(text) {
    const content = text.replace(/^---.*?---/s, '');
    return content.split(/\n\n+/).map(b => b.trim()).filter(b => b.length > 40);
}

function main() {
    const rootDir = 'content';
    const files = getMdxFiles(rootDir);
    const blocksMap = new Map();

    files.forEach(file => {
        const text = fs.readFileSync(file, 'utf8');
        const blocks = cleanText(text);
        blocks.forEach(block => {
            if (!blocksMap.has(block)) {
                blocksMap.set(block, []);
            }
            blocksMap.get(block).push(file);
        });
    });

    let found = false;
    for (const [block, occurrences] of blocksMap.entries()) {
        if (occurrences.length > 1) {
            found = true;
            console.log('--- DUPLICATE BLOCK ---');
            console.log(block);
            console.log('\nFound in:');
            console.log(occurrences.join('\n'));
            console.log('\n');
        }
    }

    if (!found) {
        console.log('No duplicate blocks found.');
    }
}

main();

File: scratch/find_duplicates.py
import os
import re
from collections import Counter

def get_mdx_files(root_dir):
    files = []
    for root, dirs, filenames in os.walk(root_dir):
        for filename in filenames:
            if filename.endswith('.mdx'):
                files.append(os.path.join(root, filename))
    return files

def clean_text(text):
    # Remove frontmatter
    text = re.sub(r'^---.*?---', '', text, flags=re.DOTALL)
    # Split into paragraphs/blocks
    blocks = [b.strip() for b in text.split('\n\n') if b.strip()]
    return blocks

def main():
    root_dir = 'content'
    files = get_mdx_files(root_dir)
    all_blocks = {}

    for file in files:
        with open(file, 'r', encoding='utf-8') as f:
            content = f.read()
            blocks = clean_text(content)
            for b in blocks:
                if len(b) < 40: continue # skip short lines
                if b not in all_blocks:
                    all_blocks[b] = []
                all_blocks[b].append(file)

    duplicates = {k: v for k, v in all_blocks.items() if len(v) > 1}
    
    if not duplicates:
        print("No duplicate blocks found.")
        return

    for block, files in duplicates.items():
        print(f"--- DUPLICATE BLOCK ---\n{block}\n\nFound in:\n" + "\n".join(files) + "\n\n")

if __name__ == "__main__":
    main()
